### PR TITLE
[MRG+1] BF: issue #570 (including tests)

### PR DESCRIPTION
**Pull request by dohmatob**

### CODE REVIEW

**Comment by dohmatob** [comment]
3 ==> 2 subjects in test.

**Comment by dohmatob** [comment]
sum ==> max heuristic

File: nilearn/decomposition/tests/test_canica.py [hunk 1]
```diff
@@ -67,5 +80,30 @@ def test_canica_square_img():
...
+    data, mask_img, components, rng = _make_canica_test_data(n_subjects=2)
+
+    # fill with -1's
+    components[:] = -1
+
+    # have some +1's (so things are not degenerate), but still have more -1's
+    # than +1's
+    for mp in components:
+        mp[rng.randn(*mp.shape) > .8] *= -.5
+        assert_less_than(mp.max(), -mp.min())
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: I don't understand the few lines above: what are they there for? It seems that they are not used.

Comment (dohmatob) [inline nilearn/decomposition/tests/test_canica.py]: Yes the code snippet is useless as it stands. Thanks for spotting. I'm pushing the full test-case (where the above snippet is used).

File: nilearn/decomposition/tests/test_canica.py [hunk 2]
```diff
@@ -3,17 +3,34 @@
...
-    shape = (20, 20, 1)
-    affine = np.eye(4)
-    rng = np.random.RandomState(0)
+try:
+    from nose.tools import assert_less_than
+except ImportError:
+    def assert_less_than(a, b):
+        if a > b:
+            raise AssertionError("%f is not less than %f" % (a, b))
+
```
Comment (lesteve) [inline nilearn/decomposition/tests/test_canica.py]: Just use assert_true I would say. If we care about providing an assert_less, it should be done somewhere in _utils rather than in one test file.

Two comments:
* assert_less_than does not exist it seems (it is called assert_less AFAICT)
* what you implemented is actually assert_less_equal

File: nilearn/decomposition/canica.py [hunk 3]
```diff
@@ -185,4 +185,10 @@ def fit(self, imgs, y=None, confounds=None):
             ica_maps[abs_ica_maps < threshold] = 0.
         self.components_ = ica_maps
 
+        # BF for issue #570: flip signs in each component to have more +ve than
+        # -ve voxels
```
Comment (lesteve) [inline nilearn/decomposition/canica.py]: I don't think this comment is accurate anymore now that we use the max heuristics.

Also I'd rather have a comment that explain that it is a heuristics for reproducible maps rather than mentioning the GH issue.

File: nilearn/decomposition/tests/test_canica.py [hunk 4]
```diff
@@ -67,5 +90,36 @@ def test_canica_square_img():
...
+    # make more de-activations than activations
+    for mp in components:
+        mp[rng.randn(*mp.shape) > .8] *= -1.
+        assert_less_than(mp.max(), -mp.min())  # goal met ?
+    data = _make_data_from_components(components, affine, shape,
+                                      rng=rng, n_subjects=2)
+
+    # synthesize data with given components
+    data = _make_data_from_components(components, affine, shape, rng=rng,
+                                      n_subjects=2)
```
Comment (lesteve) [inline nilearn/decomposition/tests/test_canica.py]: Is that not the exact same thing as l. 107-108?

File: nilearn/decomposition/tests/test_canica.py [hunk 5]
```diff
@@ -67,5 +90,36 @@ def test_canica_square_img():
...
+def test_component_sign():
+    # BF for issue #570
+
+    # make data (SVD)
+    rng = np.random.RandomState(0)
+    shape = (20, 10, 1)
+    affine = np.eye(4)
+    components = _make_canica_components(shape)
+
+    # make more de-activations than activations
```
Comment (lesteve) [inline nilearn/decomposition/tests/test_canica.py]: This comment is not accurate anymore since we are using a max heuristic.

File: nilearn/decomposition/tests/test_canica.py [hunk 6]
```diff
@@ -67,5 +90,36 @@ def test_canica_square_img():
...
+    for mp in components:
+        mp[rng.randn(*mp.shape) > .8] *= -1.
```
Comment (lesteve) [inline nilearn/decomposition/tests/test_canica.py]: In your example you have actually mp.max() == 1 == -mp.min() so the maps don't get the sign flip. If you are wondering, the reason your test pass is that your test is less or equal and it hits the  equal case in both directions.

File: nilearn/decomposition/tests/test_canica.py [hunk 7]
```diff
@@ -67,5 +90,36 @@ def test_canica_square_img():
...
+        assert_less_than(mp.max(), -mp.min())  # goal met ?
+    data = _make_data_from_components(components, affine, shape,
+                                      rng=rng, n_subjects=2)
+
+    # synthesize data with given components
+    data = _make_data_from_components(components, affine, shape, rng=rng,
+                                      n_subjects=2)
+    mask_img = nibabel.Nifti1Image(np.ones(shape, dtype=np.int8), affine)
+
+    # fit run CanICA at different times of the day
```
Comment (lesteve) [inline nilearn/decomposition/tests/test_canica.py]: This comment is not far from useless. Even if I knew what you were trying to test it took me a while to make sense of it.

File: nilearn/decomposition/tests/test_canica.py (hunk 2 above)
Comment (dohmatob) [inline nilearn/decomposition/tests/test_canica.py]: yep, there was typo in the function name/

File: nilearn/decomposition/tests/test_canica.py [hunk 8]
```diff
@@ -3,17 +3,34 @@
...
 import nibabel
-
 from nilearn.decomposition.canica import CanICA
 
 
-def test_canica_square_img():
-    shape = (20, 20, 1)
-    affine = np.eye(4)
-    rng = np.random.RandomState(0)
+try:
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: Could you please move this to a new file in nilearn/_utils/fixes/testings and mention in a comment that when we support scikit-learn 0.14 we can remove it.

File: nilearn/decomposition/tests/test_canica.py [hunk 9]
```diff
@@ -67,5 +90,34 @@ def test_canica_square_img():
...
+
+    # run CanICA many times (this is known to produce different results)
+    canica = CanICA(n_components=4, random_state=rng, mask=mask_img)
+    for _ in range(3):
+        canica.fit(data)
+        maps = canica.masker_.inverse_transform(canica.components_).get_data()
+        maps = np.rollaxis(maps, 3, 0)
+        for mp in maps:
+            assert_less_equal(-mp.min(), mp.max())
+
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: For PEP8 there should be 2 empty lines between first-level entries, thus 2 empty lines here.

File: nilearn/decomposition/tests/test_canica.py [hunk 10]
```diff
@@ -67,5 +90,34 @@ def test_canica_square_img():
...
+    data = _make_data_from_components(components, affine, shape, rng=rng,
+                                      n_subjects=2)
+    mask_img = nibabel.Nifti1Image(np.ones(shape, dtype=np.int8), affine)
+
+    # run CanICA many times (this is known to produce different results)
+    canica = CanICA(n_components=4, random_state=rng, mask=mask_img)
+    for _ in range(3):
+        canica.fit(data)
+        maps = canica.masker_.inverse_transform(canica.components_).get_data()
+        maps = np.rollaxis(maps, 3, 0)
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: It would be cooler / more readable if you used iter_img on the img rather than rollaxis on the data.

**Comment by GaelVaroquaux** [comment]
No further comments from me: should be mergeable once the above comments are addressed.

File: nilearn/decomposition/tests/test_canica.py (hunk 8 above)
Comment (banilo) [inline nilearn/decomposition/tests/test_canica.py]: cf. `assert_raises_regex` appears to be similar case and is located in nilearn/_utils/testing.py

File: nilearn/decomposition/tests/test_canica.py [hunk 11]
```diff
@@ -3,17 +3,34 @@
...
-    rng = np.random.RandomState(0)
+try:
+    from nose.tools import assert_less_equal
+except ImportError:
+    def assert_less_equal(a, b):
+        if a > b:
+            raise AssertionError("%f is not less than %f" % (a, b))
+
+
+def _make_data_from_components(components, affine, shape, rng=None,
```
Comment (banilo) [inline nilearn/decomposition/tests/test_canica.py]: Function name, even if private, could be more explicit, particularly with respect to the format of the output.

File: nilearn/decomposition/tests/test_canica.py [hunk 12]
```diff
@@ -67,5 +90,34 @@ def test_canica_square_img():
     # Smoke test to make sure an error is raised when no data is passed.
     assert_raises(TypeError, canica.fit)
 
+
+def test_component_sign():
+    # BF for issue #570
```
Comment (banilo) [inline nilearn/decomposition/tests/test_canica.py]: Mention "regression test" at this point? ...perhaps instead of "BF".

Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: > > +def test_component_sign():
> > +    # BF for issue #570

> Mention "regression test" at this point? ...perhaps instead of "BF".

Good point. And don't put a number, which doesn't mean much to the
reader, but a short description of the issue.

File: nilearn/decomposition/tests/test_canica.py (hunk 11 above)
Comment (dohmatob) [inline nilearn/decomposition/tests/test_canica.py]: Please could you create a separate issue ? This is not related to / was not introduced by the following PR.

File: nilearn/decomposition/tests/test_canica.py (hunk 9 above)
Comment (dohmatob) [inline nilearn/decomposition/tests/test_canica.py]: This recommendation is strange. My PEP8 checker dislikes it. Also, does it really matter here ? :)

**Comment by dohmatob** [comment]
OK, I've completed my end of the deal.

**Comment by dohmatob** [comment]
OK, the coveralls threshold (for acceptable % coverage decrease) seems crazy :)  Is 0.01 not simply too tiny ?

**Comment by GaelVaroquaux** [comment]
Don't worry about the coverage report.

File: nilearn/decomposition/tests/test_canica.py [hunk 13]
```diff
@@ -3,17 +3,34 @@
...
-
 from nilearn.decomposition.canica import CanICA
+from nilearn.image import iter_img
 
+try:
+    from nose.tools import assert_less_equal
+except ImportError:
+    def assert_less_equal(a, b):
+        if a > b:
+            raise AssertionError("%f is not less than %f" % (a, b))
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: Please move the above to nilearn._utils.testing

Comment (dohmatob) [inline nilearn/decomposition/tests/test_canica.py]: Done.

**Comment by GaelVaroquaux** [comment]
:+1: for merge.
